### PR TITLE
Iss1478

### PR DESCRIPTION
**Pull request by AaronJhaj**
- coping with reruns by creating a map of submitted-time: [ ]Run. This groups together reruns
- then name the folders accordingly and download artifacts
- unit tests

### CODE REVIEW

File: pkg/runs/runsDownload.go [hunk 1]
```diff
@@ -29,24 +30,75 @@ func DownloadArtifacts(
...
+			for _, run := range runs {
+				queuedTime = run.TestStructure.GetQueued()
+				runsWithSameQueuedTime := reRunsByQueuedTime[queuedTime]
+				runsWithSameQueuedTime = append(runsWithSameQueuedTime, run)
+				reRunsByQueuedTime[queuedTime] = runsWithSameQueuedTime
+			}
+
+			// SORTING BY START TIME TO DO - on api side not in cli
+
+			for _, reRunsList := range reRunsByQueuedTime {
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:51]: can this be moved to a sub-function ?

File: pkg/runs/runsDownload.go [hunk 2]
```diff
@@ -29,24 +30,75 @@ func DownloadArtifacts(
...
+				for count, reRun := range reRunsList {
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:52]: loop should break out if there is an error ?

File: pkg/runs/runsDownload.go [hunk 3]
```diff
@@ -29,24 +30,75 @@ func DownloadArtifacts(
...
+					directoryName := nameArtifactDownloadDirectory(reRun, count)
+					err = downloadArtifactsToDirectory(apiServerUrl, directoryName, reRun, fileSystem, forceDownload)
 				}
 			}
+
+		} else {
+			for _, run := range runs {
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:59]: Loop should fail on first error it gets. Else the error is over-written and lost. Loop needs to break out.

File: pkg/runs/runsDownload.go [hunk 4]
```diff
@@ -29,24 +30,75 @@ func DownloadArtifacts(
...
+				err = downloadArtifactsToDirectory(apiServerUrl, runName, run, fileSystem, forceDownload)
+			}
+			//err = downloadArtifactsToDirectory(apiServerUrl, runName, runs[0], fileSystem, forceDownload)
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:62]: Remove.

File: pkg/runs/runsDownload.go [hunk 5]
```diff
@@ -29,24 +30,75 @@ func DownloadArtifacts(
...
+func downloadArtifactsToDirectory(apiServerUrl string,
+	directoryName string,
+	reRun galasaapi.Run,
+	fileSystem utils.FileSystem,
+	forceDownload bool) error {
+	runId := reRun.GetRunId()
+	artifactPaths, err := GetArtifactPathsFromRestApi(runId, apiServerUrl)
+	if err == nil {
+		log.Printf("Creating folder %s", directoryName)
+		for _, artifactPath := range artifactPaths {
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:93]: break out the loop on an error

File: pkg/runs/runsDownload_test.go [hunk 6]
```diff
@@ -167,41 +260,61 @@ func WriteMockRasRunsFilesResponse(
...
+						for _, artifact := range artifacts {
+							if req.URL.Path == (runsFilesEndpoint + artifact.path) {
+								WriteMockRasRunsFilesResponse(t, writer, req, artifact.path)
+							}
+						}
+					}
 				}
 			}
 		}
+		// switch req.URL.Path {
```
Comment (techcobweb) [inline pkg/runs/runsDownload_test.go:300]: Remove all this commented-out code pls. It looks messy. If we don't need to run it, we don't need it.

**Review by techcobweb — CHANGES_REQUESTED** [review]
Some changes. Mostly about error messages being lost within loops.

File: pkg/runs/runsDownload.go [hunk 7]
```diff
@@ -29,24 +30,84 @@ func DownloadArtifacts(
...
+				if err == nil {
+					err = downloadArtifactsToDirectory(apiServerUrl, runName, run, fileSystem, forceDownload)
+				}
+			}
+		}
+	}
+	return err
+}
+
+func createMapOfReRuns(runs []galasaapi.Run) map[string][]galasaapi.Run {
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:64]: What is the map of ? What's the key ?

Comment (techcobweb) [reply pkg/runs/runsDownload.go:64]: eg: getRunsListpByQueuedTime ?

Doesn't this cause the queued times to be out of order, as you are putting things into a map of queued times ?

Perhaps a list of ascending queued times, with each containing a list of runs ?

**Comment by galasa-team** [comment]
Build successful

File: pkg/runs/runsDownload.go [hunk 8]
```diff
@@ -29,24 +30,84 @@ func DownloadArtifacts(
...
+						if err == nil {
+							directoryName := nameArtifactDownloadDirectory(reRun, reRunIndex)
+							err = downloadArtifactsToDirectory(apiServerUrl, directoryName, reRun, fileSystem, forceDownload)
+						}
 					}
 				}
 			}
+
+		} else {
+			for _, run := range runs {
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:54]: Surely there is only one run ? So do we need the loop ?

Comment (AaronJhaj) [reply pkg/runs/runsDownload.go:54]: agreed

File: pkg/runs/runsDownload.go (hunk 7 above)
Comment (AaronJhaj) [reply pkg/runs/runsDownload.go:64]: key = queued time
value = []runs
so then we can see a list of runs with the the same queued time.
this is helpful if we had two unrelated tests with the same runname which both had reruns (very edge-case)

File: pkg/runs/runsDownload.go [hunk 9]
```diff
@@ -29,24 +30,84 @@ func DownloadArtifacts(
...
-				for _, artifactPath := range artifactPaths {
-					var artifactData io.Reader
-					artifactData, err = GetFileFromRestApi(runId, strings.TrimPrefix(artifactPath, "/"), apiServerUrl)
-					if err == nil {
-						err = WriteArtifactToFileSystem(fileSystem, runName, artifactPath, artifactData, forceDownload)
+		if len(runs) > 1 {
+			// get list of runs that are reRuns - get list of runs that are reRuns of each other
+			// create a map of lists of reRuns - key is queued time, value is the run
+
+			reRunsByQueuedTime := createMapOfReRuns(runs)
```
[the remark this answers is not in the document]
Comment (AaronJhaj) [reply pkg/runs/runsDownload.go:40]: Which part exactly?

File: pkg/runs/runsDownload.go (hunk 7 above)
Comment (AaronJhaj) [reply pkg/runs/runsDownload.go:64]: the queued-time doesn't need to be in order. if the runs come back sorted on queued-time and there are re-runs then these will all have the same queued-time. then these re-runs should be sorted on start-time. the earliest start-time comes in and gets added the the list associated with the queued-time in the map. then the second earliest.. to the latest so the list of runs in the map will still be ordered?

File: pkg/runs/runsDownload.go [hunk 10]
```diff
@@ -29,24 +30,103 @@ func DownloadArtifacts(
...
+				// get list of runs that are reRuns - get list of runs that are reRuns of each other
+				// create a map of lists of reRuns - key is queued time, value is the run
+				reRunsByQueuedTime := createMapOfReRuns(runs)
+
+				err = downloadReRunArtfifacts(reRunsByQueuedTime, forceDownload, fileSystem, apiServerUrl)
+
+			} else if len(runs) == 1 {
+				err = downloadArtifactsToDirectory(apiServerUrl, runName, runs[0], fileSystem, forceDownload)
+			} else {
+				err = console.WriteString("No artifacts to download for run: '" + runName + "'\n")
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:50]: This message should have a GAL number associated. It should be an information message.

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful

File: pkg/runs/runsDownload.go [hunk 11]
```diff
@@ -29,19 +30,47 @@ func DownloadArtifacts(
...
+	var err error = nil
+	for _, reRunsList := range reRunsByQueuedTime {
+		if err == nil {
+			for reRunIndex, reRun := range reRunsList {
+				if err == nil {
+					directoryName := nameReRunArtifactDownloadDirectory(reRun, reRunIndex, timeService)
+					err = downloadArtifactsToDirectory(apiServerUrl, directoryName, reRun, fileSystem, forceDownload, console)
 					if err == nil {
-						err = WriteArtifactToFileSystem(fileSystem, runName, artifactPath, artifactData, forceDownload)
+						err = console.WriteString("Created folder: '" + directoryName + "'\n")
```
Comment (techcobweb) [inline pkg/runs/runsDownload.go:73]: Is there an assert statement in the unit test to check this ? Couldn't see it...

**Review by techcobweb — APPROVED** [review]
Approving, but please check that unit test question...

File: pkg/runs/runsDownload.go (hunk 11 above)
Comment (AaronJhaj) [reply pkg/runs/runsDownload.go:73]: I have on line 665 in runsDownload_test. but I will add for the other tests

**Comment by galasa-team** [comment]
Build successful

**Comment by galasa-team** [comment]
Build successful
